### ISSUE TITLE
Image for  Raspberry 400?

### ISSUE THREAD

**Issue by pwasiewi** [GenPi64/gentoo-on-rpi-64bit#8]
Has anyone tested Raspberry 400 with gentoo?
Could you put here some instructions (links?) for making image for it?

**Comment by perkinslr** [comment]
We only can support hardware we have access to.  If someone *has* one of these and is willing to test and report issues with it, we're happy to resolve any issues it has with GenPi (and the base system should "just work").

**Comment by ormorph** [comment]
>Has anyone tested Raspberry 400 with gentoo?

The Raspberry 400 is based on Pi4, so there shouldn't be any problems. The Pi4 may experience wifi issues. For a solution, you will need to install the files from [here](https://github.com/RPi-Distro/firmware-nonfree/raw/master/brcm/). You need the `brcmfmac43455-sdio.bin` file and copy it to the /lib/firmware/brcm/ directory. Failed to get this to work on `sys-kernel/linux-firmware-20201218` had to be masked.

I like the Pi4 option better, because with [this](https://www.aliexpress.com/item/32963307586.html) cooling, it is quite possible to build gentoo on a Pi4 without using a third-party PC. The 8GiB version builds even `dev-qt/qtwebengine` quite well.

**Comment by perkinslr** [comment]
> For a solution, you will need to install the files from [here](https://github.com/RPi-Distro/firmware-nonfree/raw/master/brcm/). You need the `brcmfmac43455-sdio.bin` file and copy it to the /lib/firmware/brcm/ directory. Failed to get this to work on `sys-kernel/linux-firmware-20201218` had to be masked.

In theory, the wifi ucode is available from `sys-firmware/raspberrypi-wifi-ucode::gentoo`.  It's also possible to use `savedconfig` with `linux-firmware` to install other linux firmware files, but not the 43455 broadcom stuff.  That said, on a Pi4, nothing in `linux-firmware` appears to actually be used.  That would change on a compute module with pcie-expansion stuff, but even there it's probably 50/50 on actually needing `linux-firmware`.

**Comment by ormorph** [comment]
I installed `firmware/raspberrypi-wifi-ucode`, but I had to abandon it in favor of brcmfmac43455-sdio.bin.  Wi-Fi worked, but at the time of the load on the processor it was malfunctioning.  For example, the access point disappeared from time to time. This was happening at compile time. It may be necessary to create a 9999 version of this package.

**Comment by G10h4ck** [comment]
AFAIU GenPI64 provided images boot fine on Raspberry Pi 400, while a bit of trickery is needed to make the WiFi work. Am I right?

**Comment by ormorph** [comment]
No, it was a problem when migrating from old builds from sakaki.
There are problems with btrfs in current builds. It is recommended that after writing to the USB flash drive, re-format the btrfs partition and copy the data from the image to the formatted partition.
Most likely the problem is related to the fact that the image was formed using qemu.

**Comment by G10h4ck** [comment]
So I have downloaded the image from here
https://fi.packages.genpi64.com/genpi-aarch64-desktop-latest.img.zst
The main link give 404 so I have used the mirror

if after flashing de SD i execute

`btrfstune -u /dev/MYPARTITION`

It doesn't boot anymore, basically the kernel panic because it is not able to find the root partition, so I re-flashed without altering the UUID, this way it boot without problem, but doesn't login automatically as per README, inserting demouser and the password it login without fuss.

Then in the terminal I see that both wlan' and eth' are listed, wlan0 is up (haven't tested if it really connect), eth0 is down, if I try to bring it up with `ip link set dev eth0 up` it fail and in `dmesg` some message about not being capable of finding the devie appears toghether with other bcmXXXXX messages that seems related, maybe the ethernet chip need a different driver then raspberrypi 4B ?

**Comment by G10h4ck** [comment]
I have moved the same SD to another Raspberry Pi 400 I have and there eth0 is up at boot and works fine without dmesg errors, the only difference that I know the two Pi400 have is that one is the one with spanish keyboard and the other have italian keyboard, so
PI400-es eth0 works fine with both raspberrypiOS and ubuntu, but not with genpi-aarch64-desktop-latest
PI400-it eth0 works fine with both raspberrypiOS and ubuntu, and works fine with genpi-aarch64-desktop-latest

As I expected on this kind of device `lspci -vv` doesn't output anything interesting about the ethernet card any suggestion on how further investigate what can it be the difference between the two device and contribute back?

Cheers!

**Comment by ormorph** [comment]
You also need to check the btrfs section after writing to it, for example, after updating the ports tree.
See the problem [here.](https://github.com/GenPi64/gentoo-on-rpi-64bit/issues/3#issuecomment-883887048)
>As I expected on this kind of device lspci -vv doesn't output anything interesting about the ethernet card any suggestion on how further investigate what can it be the difference between the two device and contribute back?

```
$ cat /sys/class/net/eth0/device/uevent
DRIVER=bcmgenet
OF_NAME=ethernet
OF_FULLNAME=/scb/ethernet@7d580000
OF_COMPATIBLE_0=brcm,bcm2711-genet-v5
OF_COMPATIBLE_N=1
OF_ALIAS_0=ethernet0
```

```
$ zgrep BCMGENET /proc/config.gz
CONFIG_BCMGENET=y
```
<a href="https://piccy.info/view3/15213765/f83ecc15392d846f37432d8691327dcb/1200/" target="_blank"><img src="https://i2.piccy.info/i9/95d940253e907305dbbeeb37b7f5f0ac/1642955690/74578/1398046/89629screen_trinity3_800.jpg" alt="Piccy.info - Free Image Hosting" border="0" /></a>

**Comment by G10h4ck** [comment]
This is what I get in the PI400-it with genpi-aarch64-desktop-latest that is the one which works fine

```
$ cat /sys/class/net/eth0/device/uevent
DRIVER=bcmgenet
OF_NAME=ethernet
OF_FULLNAME=/scb/ethernet@7d580000
OF_COMPATIBLE_0=brcm,bcm2711-genet-v5
OF_COMPATIBLE_N=1
OF_ALIAS_0=ethernet0
MODALIAS=of:NethernetT(null)Cbrcm,bcm2711-genet-v5
$ zgrep BCMGENET /proc/config.gz
CONFIG_BCMGENET=y
```

This is what I get in the PI400-es (the one on which eth0 doesn't come up with genpi-aarch64-desktop-latest ) with with Raspberry Pi OS

```
$ cat /sys/class/net/eth0/device/uevent
DRIVER=bcmgenet
OF_NAME=ethernet
OF_FULLNAME=/scb/ethernet@7d580000
OF_COMPATIBLE_0=brcm,bcm2711-genet-v5
OF_COMPATIBLE_N=1
OF_ALIAS_0=ethernet0
MODALIAS=of:NethernetT(null)Cbrcm,bcm2711-genet-v5

$ lsmod | grep -i bcm
btbcm                  16384  1 hci_uart
bluetooth             393216  31 hci_uart,bnep,btbcm,rfcomm
bcm2835_codec          40960  0
bcm2835_v4l2           40960  0
v4l2_mem2mem           36864  1 bcm2835_codec
bcm2835_isp            32768  0
bcm2835_mmal_vchiq     32768  3 bcm2835_isp,bcm2835_codec,bcm2835_v4l2
videobuf2_vmalloc      16384  1 bcm2835_v4l2
videobuf2_dma_contig    20480  2 bcm2835_isp,bcm2835_codec
videobuf2_v4l2         32768  4 bcm2835_isp,bcm2835_codec,bcm2835_v4l2,v4l2_mem2mem
videobuf2_common       61440  5 bcm2835_isp,bcm2835_codec,bcm2835_v4l2,v4l2_mem2mem,videobuf2_v4l2
videodev              249856  6 bcm2835_isp,bcm2835_codec,videobuf2_common,bcm2835_v4l2,v4l2_mem2mem,videobuf2_v4l2
mc                     45056  7 bcm2835_isp,bcm2835_codec,snd_usb_audio,videobuf2_common,videodev,v4l2_mem2mem,videobuf2_v4l2
snd_bcm2835            24576  0
vc_sm_cma              32768  6 bcm2835_isp,bcm2835_mmal_vchiq
snd_pcm               110592  6 snd_compress,snd_usb_audio,snd_pcm_dmaengine,snd_soc_hdmi_codec,snd_bcm2835,snd_soc_core
snd                    77824  17 snd_compress,snd_hwdep,snd_usb_audio,snd_soc_hdmi_codec,snd_timer,snd_rawmidi,snd_usbmidi_lib,snd_seq_device,snd_bcm2835,snd_soc_core,snd_pcm
```

no `/proc/config.gz` there I hope the modules list is enough

AFAIU can't see sensible differences in the outputs :-[

**Comment by ormorph** [comment]
>no /proc/config.gz there I hope the modules list is enough

Most likely the kernel config is in the /boot directory.

What does the command show?:
```
$ dmesg |grep -i err
```
and:
```
$ dmesg |grep -i eth
```
This is on PI400-es -> genpi-aarch64-desktop-latest.
